andela/ah-the-jedi-backend
# #165305276 Add Search and Filter for articles

### What does this PR do?
Adds search and filter functionality for articles

### Description of Task(s) to be completed?
-Users should be able to search for articles by author, article title and tag/keyword or search by filtering.
-Users should be able to filter articles by author, article title and tag/keywords. 

### How should this be manually tested?

Clone this repository on local machine and CD into project folder
`$ git clone git@github.com:andela/ah-the-jedi-backend.git`
`$ cd ah-the-jedi-backend`

Check out to the PR's branch
`$ git checkout ft-search-filter-165305276`

Create and activate virtual environment
`$ python3 -m venv venv`
`$ source .venv/bin/activate`

Install app dependencies from requirements.txt
`$ pip install -r api/requirements.txt`

Make migrations and run the server
`$ python manage.py makemigrations`
`$ python manage.py migrate`
`$ python manage.py runserver`

**Use postman to test the endpoints**


           POST <your-localhost>/api/users/



       POST <your-localhost>/api/login/ 

       Use the token provided after login for authorization 

- [x] **Create an article**

      `POST`  `<you-local-host>/api/articles/`

- [x] **Filter by author**

      `GET`  `<you-local-host>/api/article/search/?author=<author>`

### Screenshot
<img width="942" alt="Screenshot 2019-05-08 at 17 27 10" src="https://user-images.githubusercontent.com/45200891/57383103-e2871180-71b6-11e9-91e9-374adb5cc0b1.png">




- [x] **Filter by title**

      `GET`  `<you-local-host>/api/article/search/?title=<title>`

### Screenshot
<img width="941" alt="Screenshot 2019-05-08 at 17 26 42" src="https://user-images.githubusercontent.com/45200891/57383132-efa40080-71b6-11e9-8f72-77e98761f8f9.png">




- [x] **Filter by tag**

      `GET`  `<you-local-host>/api/article/search/?tag=<tag>`

### Screenshot
<img width="940" alt="Screenshot 2019-05-08 at 17 27 35" src="https://user-images.githubusercontent.com/45200891/57383162-fb8fc280-71b6-11e9-84b6-ae1ba68350e2.png">





- [x] **Filter by multiple parameters**

      `GET`  `<you-local-host>/api/article/search/?author=<author>&title=<title>&tag=<tag>`

### Screenshot
<img width="941" alt="Screenshot 2019-05-08 at 17 28 11" src="https://user-images.githubusercontent.com/45200891/57383202-09ddde80-71b7-11e9-9db7-1dbca0252a01.png">






      `GET`  `<you-local-host>/api/article/search/?search=<search-parameter>`

### Screenshot
<img width="943" alt="Screenshot 2019-05-08 at 17 28 35" src="https://user-images.githubusercontent.com/45200891/57383225-12ceb000-71b7-11e9-82b5-7e431b393cef.png">





### What are the relevant PT stories?
[#165305276](https://www.pivotaltracker.com/story/show/165305276)
